src/frontend/src/scheduler/distributed/query.rs

```diff
@@ -206,80 +214,68 @@ impl QueryRunner {
         let mut stages_with_table_scan = self.query.stages_with_table_scan();
 
         // Schedule other stages after leaf stages are all scheduled.
-        while let Some(msg) = self.msg_receiver.recv().await {
-            match msg {
-                Stage(Scheduled(stage_id)) => {
-                    tracing::trace!(
-                        "Query stage {:?}-{:?} scheduled.",
-                        self.query.query_id,
-                        stage_id
-                    );
-                    self.scheduled_stages_count += 1;
-                    stages_with_table_scan.remove(&stage_id);
-                    if stages_with_table_scan.is_empty() {
-                        // We can be sure here that all the Hummock iterators have been created,
-                        // thus they all successfully pinned a HummockVersion.
-                        // So we can now unpin their epoch.
-                        tracing::trace!("Query {:?} has scheduled all of its stages that have table scan (iterator creation).", self.query.query_id);
-                        self.hummock_snapshot_manager
-                            .unpin_snapshot(self.epoch, self.query.query_id())
-                            .await;
-                    }
+        let mut shutdown_rx = shutdown_rx_for_cancel;
+        loop {
+            tokio::select! {
+                _ = &mut shutdown_rx => {
 
-                    if self.scheduled_stages_count == self.stage_executions.len() {
-                        // Now all stages have been scheduled, send root stage info.
-                        self.send_root_stage_info().await;
-                    } else {
-                        for parent in self.query.get_parents(&stage_id) {
-                            if self.all_children_scheduled(parent).await
-                                // Do not schedule same stage twice.
-                                && self.stage_executions[parent].is_pending().await
-                            {
-                                self.stage_executions[parent].start().await;
-                            }
-                        }
-                    }
+                    self.handle_cancel_or_failed_stage(shutdown_tx, SchedulerError::QueryCancelError).await;
```

Comment:
Why not just add an `CancelQuery` msg, this way we don't need another shutdown_tx

Reply:
Good point. I checked the code found that we can achieve that by passing some arguments. 

Some things not change:
* Still need to have a Map in context (query_id -> channel sender)
* The CancelQueryMsg may be blocked by other message (Scheduled), may not be so effectively for cancel.

Reply:
Oops, seems I have misunderstand some points. Let me reorganize my words

Reply:
Note that `shutdown_tx` is used for sending shutdown message to QueryResultFetcher, and not introduced by this PR but #5022.

For the question, yes, I think we can do that. But the problem is we might need to move the `channel::new` into `handle_query`, this is where the shutdown is inited. To avoid this we can do some refactor, consume `stream<DataChunk>` -> `vec<DataChunk>` (L69 - L71) in a function of `QueryResultFetcher`.  I think that will be more clear and we can do it next PR.

Reply:
Emmmm. We can not re-use the channel receiver of QueryRunner, cuz mpsc only allow one consumer, and there is already one in `QueryRunner`.

Reply:
See above.

Reply:
Add a QueryMessage::CancelQuery